File: blind-75/merge_k_sorted_lists_01.py
# leetcode 23

from typing import List, Optional
from heapq import heappush, heappop

class ListNode:
	def __init__(self, val = 0, next = None ):
		self.val = val
		self.next = next

class Solution:
	def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
		minHeap = []

		# include each first node of the list to heap
		for i, node in enumerate(lists):
			if node:
				heappush(minHeap, (node.val, i, node))

		dummy = ListNode()

		curr = dummy

		while(len(minHeap) > 0):
			# pop the top info of the heap that represents the smallest node
			topNode = heappop(minHeap)
			# print(f"node val: {topNode[0]}")

			# include the smallest node to the final list
			curr.next = ListNode(topNode[0])
			curr = curr.next

			# if the top of the heap has next node then set that node to the next
			if topNode[2].next:
				heappush(minHeap, (topNode[2].next.val, topNode[1], topNode[2].next))

		return dummy.next



obj = Solution()

lists = [
	ListNode(1, ListNode(4, ListNode(5))),
	ListNode(1, ListNode(3, ListNode(4))),
	ListNode(2, ListNode(6))
]

obj.mergeKLists(lists)

File: blind-75/merge_k_sorted_lists_01.java
// leetcode 23

import java.util.*;

class ListNode {
	int val;
	ListNode next;
	ListNode () {}
	ListNode (int val) {
		this.val = val;
	}
	ListNode (int val, ListNode next) {
		this.val = val;
		this.next = next;
	}
}

class NodeInfo implements Comparable<NodeInfo> {
	int val;
	int index;
	ListNode node;

	NodeInfo (int val, int index, ListNode node) {
		this.val = val;
		this.index = index;
		this.node = node;
	}

	@Override 
	public int compareTo(NodeInfo other) {
		if (this.val == other.val) {
			return Integer.compare(this.index, other.index);
		}
		return Integer.compare(this.val, other.val);
	}

	public String toString () {
		return "(val: " + this.val + ", index: " + this.index + ", node: " + this.node + ")";
	}
}

class Solution {
	public ListNode mergeKLists(ListNode[] lists) {

		ListNode dummy = new ListNode();
		PriorityQueue<NodeInfo> minHeap = new PriorityQueue<>();

		for (int i = 0; i < lists.length; i++) {
			ListNode node = lists[i];
			if (node != null) {
				minHeap.add(new NodeInfo(node.val, i, node));
			}
		}

		ListNode curr = dummy;
		while (!minHeap.isEmpty()) {
			// pop the top node
			NodeInfo topNode = minHeap.poll();
			System.out.println("node: " + topNode.val + " ");
			// include this topnode to the final answer list
			curr.next = new ListNode(topNode.val);
			curr = curr.next;
			// now add the next node of the topnode to the min heap
			if (topNode.node.next != null) {
				minHeap.add(new NodeInfo(topNode.node.next.val, topNode.index, topNode.node.next));
			}
		}

		return dummy.next;
	}
}

class Main {
	public static void main(String[] args) {
		Solution obj = new Solution();

		ListNode[] lists = new ListNode[] {
			new ListNode(1, new ListNode(4, new ListNode(5))),
			new ListNode(1, new ListNode(3, new ListNode(4))),
			new ListNode(2, new ListNode(6))
		};

		System.out.println(obj.mergeKLists(lists));
	}
}